src/modules/wave/WaveModule.v3

```diff
@@ -62,6 +68,46 @@ class WaveInstance {
 	}
 
 	// ==={Provided functions}============================================
+	def new_func(sig_idx: i32, ptr: i32, len: i32) -> i32 {
+		/* get type signature */
+		def decl = instance.heaptypes[sig_idx];
+		if (!SigDecl.?(decl))
+			return i32.view(-1);
+		def sig = SigDecl.!(decl);
+
+		/* copy from memory */
+		def result: MaybeTrap<Array<byte>> = copyString(u32.view(ptr), u32.view(len));
+		if (result.trapped())
+			return i32.view(-1);
+		def code: Array<byte> = result.result;
+
+		var func: FuncDecl = FuncDecl.new(sig_idx);
+		func.setOrigCode(code);
+		func.sig = sig;
+		func.func_index = instance.functions.length; // but not yet added
+
+		/* validate */
+		var err: ErrorGen = ErrorGen.new(instance.module.filename);
+		Execute.tiering.onFuncBody(instance.module, u32.!(func.func_index), code, err);
+		var validator = CodeValidator.new(Extension.set.all, Limits.new(), instance.module, err);
```

Comment:
Let's cache a lazily-created code validator in the WaveInstance to avoid constructing one each time.